posts/airflow-ray.md

```diff
@@ -0,0 +1,211 @@
+    evallist = [(validation_df, 'eval')]
+    evals_result = {}
+    config = {
+        "tree_method": "hist",
+        "eval_metric": ["logloss", "error"],
+    }
+    bst = xgb.train(
+        params=config,
+        dtrain=train_df,
+        evals_result=evals_result,
+        ray_params=xgb.RayParams(max_actor_restarts=1, num_actors=8, cpus_per_actor=2),
+        num_boost_round=100,
+        evals=evallist)
+    return bst
+```
+
+4. Create a DAG function with the logical flow of data
+
+```python
+@dag(default_args=default_args, schedule_interval=None, start_date=days_ago(2), tags=['finished-modin-example'])
+def task_flow_xgboost_modin():
+    build_raw_df = load_dataframe()
+    data = create_data(build_raw_df)
+    trained_model = train_model(data)
+
+
+task_flow_xgboost_modin = task_flow_xgboost_modin()
+```
+
+
+5. Upload the DAG to Airflow, and your data pipeline is now a running Airflow DAG! 
+
+With maybe 20 minutes of work, a data scientist can turn a local Python script into a high-scale, reproducible pipeline with the power of Ray distributed computing and Airflow orchestration. These data engineers can then take advantage of Airflow [variables](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#variables), [connections](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#connections), scheduling intervals, and many other features that have made Airflow so ubiquitous with production-grade scheduling and orchestration.
+
+
+### Passing data between tasks: faster with Plasma
+
+One thing that experienced Airflow users will notice in the example above is that we appear to be passing entire dataframes between tasks without explicitly sending those data chunks to external storage. With traditional XCom, this would be essentially impossible because Airflow stores each piece of data sent between tasks in a single cell of the metadata DB. 
+
+To address this issue, we get to take advantage of one of Ray’s coolest features: data caching. To ensure fast data processing for ML, Ray utilizes a_ plasma store_ system that caches all data in memory on Rray workers. With the Ray decorator, Airflow only stores a hash pointing to the result of the last task (a Ray Object ID). This caching system allows for data to stay in Ray between tasks without ever leaving the RAM of the workers. No more writing and reading data from S3 between tasks!
+
+While this alpha release implements the Ray plasma store for passing data between Ray tasks. Future releases will simplify moving data in and out of Ray, and possibly even extend the Ray custom Xcom Backend for moving data between all Airflow tasks.
+
+
+### Current Limitations
+
+The Ray Airflow Task API as shown above is currently in alpha, which means there will be some rough edges. We welcome any and all bug reports, suggestions, or PRs! You can find the code &lt;LINK HERE> and &lt;LINK HERE>.
+
+
+
+## More advanced usage: fault tolerance
+
+
```

Comment:
### Hyperparameter Tuning

Airflow has templating and dynamic parameterization capabilities, and when combined with [Ray Tune](https://docs.ray.io/en/master/tune/index.html), one can orchestrate and dynamically scale tuning an ML search space using any machine learning framework - including PyTorch, XGBoost, MXNet, and Keras - while easily integrating tools for recording, querying, and replicating experiments, as well as register & deploy the resulting models.

An example using XGBoost and Tune might begin like:

```python
@ray_task(**task_args)
def tune_model(data):

    search_space = {
        # You can mix constants with search space objects.
        "objective": "binary:logistic",
        "eval_metric": ["logloss", "error"],
        "max_depth": tune.randint(1, 9),
        "min_child_weight": tune.choice([1, 2, 3]),
        "subsample": tune.uniform(0.5, 1.0),
        "eta": tune.loguniform(1e-4, 1e-1),
    }

    print("enabling aggressive early stopping of bad trials")
    # This will enable aggressive early stopping of bad trials.
    scheduler = ASHAScheduler(
        max_t=10, grace_period=1, reduction_factor=2  # 10 training iterations
    )

    print("Tuning")

    analysis = tune.run(
        tune.with_parameters(train_model, data=data),
        metric="eval-logloss",
        mode="min",
        local_dir=LOCAL_DIR,
        # You can add "gpu": 0.1 to allocate GPUs
        resources_per_trial=XGB_RAY_PARAMS.get_tune_resources(),
        config=search_space,
        num_samples=10,
        scheduler=scheduler,
    )

    print("Done Tuning")

    return analysis

@ray_task(**task_args)
def load_best_model_checkpoint(analysis):

    print("Checking Analysis")

    best_bst = xgb.Booster()

    print(f"Analysis Best Result on eval-error is: {analysis.best_result['eval-error']}")
    print("Loading Model with Best Params")

    best_bst.load_model(os.path.join(
        analysis.best_checkpoint, "model.xgb"))
    accuracy = 1. - analysis.best_result["eval-error"]

    print(f"Best model parameters: {analysis.best_config}")
    print(f"Best model total accuracy: {accuracy:.4f}")

    # We could now do further predictions with
    # best_bst.predict(...)
    return best_bst

analysis = tune_model(data)
best_checkpoint = load_best_model_checkpoint(analysis)
```